### PR TITLE
Station records

### PR DESCRIPTION
**Pull request by vulppine**
## About the PR 

Adds station records, which are primarily for storing user-facing information about a player character in a central database.

Records are indexed by key, which is created the moment a general record is first created (TBD, may be modified in the future) - this can stored be in an in-game entity, which can then be used by other in-game entities to modify/access the station record entries by the stored key.

Records themselves are stored on the station virtual entity, rather than any in-game entity - a future implementer can, however, implement some kind of bridge entity for accessing these records for a more realistic approach. Other systems can interact with the primary StationRecordsSystem in order to add their own special type of record into the main set of records.

**TODO**

*Required*
- [x] Add a crew manifest listing
- [x] Add the ability to ~~modify~~ see general records from a console
- [x] Have the ID card modification computer modify crew records passively (i.e., selecting an existing job when selecting what type of job a player's character can have, changing job titles, etc)

*Maybe*
- [ ] Add the other two types of records (medical, security) - this is a maybe because it then requires those to have their own systems implemented, which may be out of scope

:cl:
- add: NanoTrasen has finally bought a license for SpaceDB, and is sharing it with its stations for crew record management. We only have the one license. Don't tell them what we're doing.

### CODE REVIEW

**Review by moonheart08 — DISMISSED** [review]

**Comment by juliangiebel** [comment]
Respawning a player as admin causes the manifest entry to duplicate. I don't know if that's something you want to fix or not but I don't see a problem with that.
image

**Review by juliangiebel — COMMENTED** [review]
Code looks good.
If an ID card is destroyed is the manifest entry supposed to vanish or not?
What about blank ID cards?

The job preset sets the translation key as the job name instead of the name
image

**Comment by vulppine** [comment]
> Respawning a player as admin causes the manifest entry to duplicate. I don't know if that's something you want to fix or not but I don't see a problem with that. image

if they join in from the lobby, it's essentially the same as joining as a new character - i don't think this should be fixed imo

**Comment by vulppine** [comment]
> Code looks good. If an ID card is destroyed is the manifest entry supposed to vanish or not? What about blank ID cards?
> 
> The job preset sets the translation key as the job name instead of the name image

right now, blank ID cards won't link to a station record, but the preset setting the job name to the localization string name is an error so

File: Content.Server/StationRecords/Systems/GeneralStationRecordConsoleSystem.cs [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+            _userInterface.GetUiOrNull(uid, GeneralStationRecordConsoleKey.Key)?.SetState(new GeneralStationRecordConsoleState(null, null, null));
+            return;
+        }
+
+        var listing = new Dictionary<StationRecordKey, string>();
+        foreach (var pair in enumerator)
+        {
+            if (pair == null)
+            {
+                return;
```
Comment (metalgearsloth) [inline Content.Server/StationRecords/Systems/GeneralStationRecordConsoleSystem.cs:63]: continue?

File: Content.Server/CrewManifest/CrewManifestSystem.cs [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+        if (shell.Player == null || shell.Player is not IPlayerSession session)
+        {
+            shell.WriteLine("You must run this from a client.");
+            return;
+        }
+
+        var crewManifestSystem = entMan.EntitySysManager.GetEntitySystem<CrewManifestSystem>();
+
+        crewManifestSystem.OpenEui(uid, session);
+    }
```
Comment (metalgearsloth) [inline Content.Server/CrewManifest/CrewManifestSystem.cs:271]: Should probably be localised I guess.

File: Content.Server/CrewManifest/CrewManifestSystem.cs [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+    [Dependency] private readonly EuiManager _euiManager = default!;
+    [Dependency] private readonly IConfigurationManager _configManager = default!;
+    [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+
+    /// <summary>
+    ///     Cached crew manifest entries. The alternative is to outright
+    ///     rebuild the crew manifest every time the state is requested:
+    ///     this is inefficient.
+    /// </summary>
+    private readonly Dictionary<EntityUid, CrewManifestEntries> _cachedEntries = new();
```
Comment (metalgearsloth) [inline Content.Server/CrewManifest/CrewManifestSystem.cs:36]: Are these cleared?

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Server/CrewManifest/CrewManifestSystem.cs (hunk 2 above)
Comment (vulppine) [reply Content.Server/CrewManifest/CrewManifestSystem.cs:272]: i feel like a lot of these are generic enough that commands could just have localization strings that aren't specific to one command - might be for another PR unless these already exist?

**Review by metalgearsloth — APPROVED** [review]

**Comment by metalgearsloth** [comment]
Testmerge time

**Pull request merged**